README
ElasticSearch.pm is a Perl API to ElasticSearch:

ElasticSearch is an Open Source (Apache 2 license), distributed, RESTful
Search Engine based on Lucene, and built for the cloud, with a JSON API.

Check out its features: L<http://www.elasticsearch.com/products/elasticsearch/>

This module is a thin API which makes it easy to communicate with an
ElasticSearch cluster.

It maintains a list of all servers/nodes in the ElasticSearch cluster, and
spreads the load randomly across these nodes.  If the current active node
disappears, then it attempts to connect to another node in the list.

Forking a process triggers a server list refresh, and a new connection to
a randomly chosen node in the list.


INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc ElasticSearch

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=ElasticSearch

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/ElasticSearch

    CPAN Ratings
        http://cpanratings.perl.org/d/ElasticSearch

    Search CPAN
        http://search.cpan.org/dist/ElasticSearch/


LICENSE AND COPYRIGHT

Copyright (C) 2010 Clinton Gormley

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for ElasticSearch
0.21    2010-09-18
        Supports ElasticSearch version 0.10.0
        Added the _source parameter to put_mapping()

0.20    2010-09-04
        Supports ElasticSearch version 0.10.0
        Added update_index_settings() plus test
        Added pod coverage test

0.19    2010-08-19
        Supports ElasticSearch version 0.9.1
        Minor bug fix to cluster_state
        Updated tests to use facet_filter instead of filter

0.18    2010-08-01
        Updated Changes file

0.17    2010-08-01

        Supports ElasticSearch version 0.9, which removes support for some
        features of older versions.

      - In search() added support for 'script_fields'
      - In count() and delete_by_query() added support for 'custom_score',
        'fuzzy', 'span_term', 'span_first', 'span_near', 'span_not',
        'span_or'
      - Removed terms() - now handled by search facets
      - Added the 'full' parameter to flush_index()
      - Changed put_mapping() to automatically prefix the type properties
        definition with the type name
      - Removed get_mapping() and replaced it with mapping(), which uses
        the newly exposed GET _mapping API
      - Added current_server_version()
      - Added the 'filter_nodes', 'filter_metadata', 'filter_routing_table'
        and 'filter_indices' params to cluster_state()



0.16    2010-05-20
        Added an ElasticSearch::Error::Timeout error class, to catch
        occassions when the request times out, instead of failing
        for a different reason.

        Also, catch 'Server closed connection' errors, and automatically
        refresh_servers and retry the request.

0.15    2010-05-16
        Fixed a bug in refresh_servers() introduced in 0.14, when forking
        processes.

0.14    2010-05-16
        The object now stores the original servers parameter, so that
        if the nodes disassociate and the live server list is limited,
        the object can still try the original servers.

0.13    2010-05-13
        Added nodes_stats(), shutdown(), restart() and camel_case()
        Removed Alien::ElasticSearch, as it is now easier to install
        ElasticSearch directly.
        Tested against ElasticSearch version 0.7.0

0.12    2010-04-16
        Added scroll() and the indices_boost param to search()
        Tested against ElasticSearch version 0.6.0

0.11    2010-04-07
        Updated changelog

0.10    2010-04-07
        Fixed the duration regex
        Improved docs for shutdown()

0.09    2010-04-07
        Changed refresh_servers to handle http_address and httpAddress
        Added a terms filter test

0.08    2010-04-05
      - Changed all_field to _all

0.07    2010-04-05

      - All tests pass as of version 8cfa04cea3bc2317ee613384ced4da20cf2737aa
        on 2010-04-04:
        http://github.com/elasticsearch/elasticsearch/commit/8cfa04cea3bc2317ee613384ced4da20cf2737aa

      - Updated the API to take into account ElasticSearch's change from
        using camelCase to underscore_separators
      - create() now uses the /_create postfix instead of op_type=create,
        although that flag is still available through index() and set()
      - Changed more_like_this and more_like_this_field to mlt and mlt_field
      - Changed fuzzy_like_this and fuzzy_like_this_field to flt and flt_field
      - Changed term() to use gt/gte/lt/lte instead of from/to and
        fromInclusive/toInclusive
      - put_mapping now defaults to { ignore_conflicts => 0 }, so the
        default now is to complain
      - In mlt, boost_terms:boolean and boost_terms_factor:float have now
        been merged to boost_terms:float

0.06    2010-04-03
      - The test suite now requires Alien::ElasticSearch, and uses its own
        config file and work dir, so as not to interfere with existing
        data
      - Added clear_cache()
      - Added fuzzy_like_this and fuzzy_like_this_field clauses to
        count() and delete_by_query()
      - Added more_like_this()
      - Added more_like_this/more_like_this_field clauses to count()
        and delete_by_query()
      - Changed exclude_from and exclude_to to from_inclusive and to_inclusive
        in the terms() query
      - the filteredQuery clause renamed to 'filtered' as per a change in ES
      - Added aliases() and get_aliases()
      - Deleted the QueryDSL doc, as the ElasticSearch docs are much improved

0.05    2010-03-04
      - Delayed connecting to the cluster until actually required (previously
        it refreshed the server list in new() )
      - Restructured the internals again, to make it easier to read
      - Added cluster_health()
      - Fixed the test suite so that auto-started nodes don't take over
        the terminal
      - Enabled more tests as ElasticSearch now less buggy

0.04    2010-02-23
      - Added terms()
      - Added get_mapping()
      - Renamed create_mapping() to put_mapping()
      - Added ignore_duplicate param to put_mapping
      - skip create-duplicate-mapping test if cluster has more than one nodes

0.03    2010-02-21
      - Corrected the docs for search/count/delete_by_query
      - Added POD for the Query DSL
      - Added a SIG{INT} to the test file, so that it shuts down the servers
        if the test suite is interrupted

0.02    2010-02-21
      - Rewrote the module completely to use a dispatch table -
        makes it easier to extend
      - Added a test suite
      - Added Alien::ElasticSearch so that we can test against a live server

0.01    2010-02-16
        First version, released on an unsuspecting world.



Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.3607
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'ElasticSearch',
          'VERSION_FROM' => 'lib/ElasticSearch.pm',
          'PREREQ_PM' => {
                           'Data::Dumper' => 0,
                           'Encode' => 0,
                           'File::Temp' => '0.22',
                           'HTTP::Request' => 0,
                           'IO::Socket' => 0,
                           'JSON::XS' => 0,
                           'LWP::ConnCache' => 0,
                           'LWP::UserAgent' => 0,
                           'POSIX' => 0,
                           'Test::Most' => 0
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [],
          'PL_FILES' => {}
        )
;


lib/ElasticSearch/AnyEvent.pm
package ElasticSearch::AnyEvent;

use strict;
use warnings FATAL => 'all';
use base 'ElasticSearch';
use Scalar::Util qw(weaken);
use AnyEvent();
use AnyEvent::HTTP qw(http_request);

our $VERSION = '0.02';

=item C<bulk_create()>

=cut

#===================================
sub bulk_create {
#===================================
    my ( $self, $params ) = ElasticSearch::_params(@_);
    return $self->bulk_index(
        action => 'create',
        %$params
    );
}

=item C<bulk_insert()>

=cut

#===================================
sub bulk_index {
#===================================
    my ( $self, $params ) = ElasticSearch::_params(@_);
    my %retried;
    return $self->bulk_action(
        action   => 'index',
        on_error => sub { _retry_action( \%retried, 'index', @_ ) },
        %$params
    );
}

=item C<bulk_delete()>

=cut

#===================================
sub bulk_delete {
#===================================
    my ( $self, $params ) = ElasticSearch::_params(@_);
    my %retried;
    return $self->bulk_action(
        action   => 'delete',
        on_error => sub { _retry_action( \%retried, undef, @_ ) },
        %$params
    );
}

=item C<_retry_action()>

=cut

#===================================
sub _retry_action {
#===================================
    my $retried    = shift;
    my $new_action = shift;
    my $error      = shift;
    if ( $error->isa('ElasticSearch::AnyEvent::Error::Timeout') ) {
        my $args = shift;
        unless ( $retried->{ $args->{id} }++ ) {
            $args->{action} = $new_action if $new_action;
            my $queue = shift;
            push @$queue, $args;
            return;
        }
    }
    warn $error;
}

=item C<bulk_search()>

=cut

#===================================
sub bulk_search {
#===================================
    my ( $self, $params ) = ElasticSearch::_params(@_);
    return $self->bulk_action(
        action => 'search',
        %$params
    );
}

#===================================
sub bulk_action {
#===================================
    my ( $self, $params ) = ElasticSearch::_params(@_);

    my $action = $params->{action}
        or $self->throw( 'Param', "Missing required param 'action'" );

    $self->throw( 'Param', "Unknown action '$action'" )
        unless $self->can($action);

    my $max       = $params->{max}        || 10;
    my $low_queue = $params->{low_queue} || $max * 2;

    my @queue        = ( @{ $params->{queue} || [] } );
    my $on_low_queue = $params->{on_low_queue};
    my $on_success   = $params->{on_success};
    my $on_error
        = exists $params->{on_error}
        ? $params->{on_error}
        : sub { warn $_[0] };

    my ( $queue_empty, $running, $errors, $group_ended ) = (0) x 4;
    $queue_empty = 1 unless $on_low_queue;

    my ( $issue_job, $croak_cb, %jobs );

    my $group_cv = $self->cv( on_destroy => sub { %jobs = () } );

    my $weak_group_cv = $group_cv;
    weaken $weak_group_cv;
    $group_cv->guard($group_cv) unless defined wantarray;

    $group_cv->begin(
        sub {
            my $cv = shift;
            $cv->send( $errors == 0 );
            $cv->clear_guard;
        }
    );

    $croak_cb = sub {
        %jobs = ();
        $weak_group_cv->croak(@_);
        $weak_group_cv->clear_guard if $weak_group_cv;
        0;
    };

    $issue_job = sub {
        return unless $running < $max;
        if ( !$queue_empty and @queue < $low_queue ) {
            my @new_args;
            eval { @new_args = $on_low_queue->( \@queue ); 1; }
                or return $croak_cb->($@);
            $queue_empty = 1 unless @new_args;
            return $new_args[0]->cb($issue_job)
                if @new_args == 1 && eval { $new_args[0]->can('cb') };
            push @queue, @new_args;
        }
        unless (@queue) {
            $group_ended++ || $weak_group_cv->end;
            return;
        }

        my $args      = shift @queue;
        my $my_action = delete $args->{action} || $action;
        my $job_cv    = eval { $self->$my_action($args) }
            or return $croak_cb->($@);
        $jobs{"$job_cv"} = $job_cv;
        weaken $job_cv;
        $running++;
        $weak_group_cv->begin;
        $job_cv->cb(
            sub {
                $running--;
                $weak_group_cv->end;
                delete $jobs{"$job_cv"};

                my $error = $@;
                eval {
                    if ( my $result = shift )
                    {
                        $on_success->( $args, $result ) if $on_success;
                    }
                    else {
                        $errors++;
                        $on_error->( $error, $args, \@queue ) if $on_error;
                    }
                    1;
                } or return $croak_cb->($@);
                $issue_job->();
            },
            'weak'
        );

        return 1;
    };
    while (1) { $issue_job->() || last }
    return $group_cv

}

#===================================
sub request {
#===================================
    my $self   = shift;
    my $params = shift;
    my $method = $params->{method} || 'GET';
    my $cmd    = $params->{cmd}
        || $self->throw( 'Request', "No cmd specified" );

    my $data = $params->{data};
    my $json = $self->JSON;

    if ( defined $data ) {
        $data = $json->encode($data);
        utf8::upgrade($data);
        utf8::encode($data);
    }

    my $cv = $self->cv;

    my $cv_weak = $cv;
    weaken $cv_weak;

    $cv->guard($cv) unless defined wantarray();
    my ( $request_cb, $server_cb, $server );

    $server_cb = sub {
        $server = shift;
        unless ($server) {
            $cv_weak->croak($@);
            return;
        }
        $cv_weak->guard(
            http_request(
                $method => $server . $cmd,
                body    => $data,
                timeout => $self->timeout,
                $request_cb
            )
        );
    };

    $request_cb = sub {
        $self->_log_request( $method, $server, $cmd, $data );
        my ( $result, $error ) = $self->_parse_response( $server, @_ );
        if ($result) {
            $cv_weak->send($result);
            $cv_weak->clear_guard if $cv_weak;
            return;
        }

        if ( ref $error
            && $error->isa('ElasticSearch::AnyEvent::Error::Connection') )
        {
            my $local_cv = $cv_weak;
            $cv_weak->clear_guard;
            $cv_weak->guard( $self->refresh_servers->cb($server_cb) );
            return;
        }
        $cv_weak->croak($error);
        $cv_weak->clear_guard if $cv_weak;
    };

    my $request = $self->current_server;
    $cv->guard($request);
    $request->cb( $server_cb, 'weak' );
    return $cv;

}

#===================================
sub current_server {
#===================================
    my $self = shift;
    if (@_) {
        $self->_set_current_server(@_);
    }

    my $cv = $self->cv;
    if ( my $current_server = $self->_get_current_server ) {
        if ( ( my @live_servers = @{ $self->servers } ) > 1 ) {
            my $next_server = shift @live_servers;
            $self->servers( [ @live_servers, $next_server ] );
            $self->_set_current_server($next_server);
        }
        $cv->send($current_server);
        return $cv;
    }

    my $refresh_cv = $self->refresh_servers;
    $cv->guard($refresh_cv);
    $cv->guard($cv) unless defined wantarray;

    my $cv_weak = $cv;
    weaken $cv_weak;

    $refresh_cv->cb(
        sub {
            $cv_weak->clear_guard;
            if ( my $current_server = shift ) {
                return $cv_weak->send($current_server);
            }
            return $cv_weak->croak($@);
        },
        'weak'
    );

    return $cv;
}

#===================================
sub refresh_servers {
#===================================
    my $self = shift;
    $self->_set_current_server(undef);

    my %servers = map { $_ => 1 }
        ( @{ $self->servers }, @{ $self->default_servers } );
    my @all_servers = keys %servers;

    my $cv = $self->cv( on_destroy => sub { %servers = () } );

    my $cv_weak = $cv;
    weaken $cv_weak;
    $cv->guard($cv) unless defined wantarray;

    foreach my $server ( keys %servers ) {
        $servers{$server} = http_request(
            GET     => $server . '/_cluster/nodes',
            timeout => $self->timeout,
            sub {
                if ( my $current = $self->_get_current_server ) {
                    %servers = ();
                    $cv_weak->send($current);
                    $cv_weak->clear_guard if $cv_weak;
                    return;
                }
                delete $servers{$server};

                $self->_log_request( 'GET', $server, '/_cluster/nodes' );
                my ( $result, $error )
                    = $self->_parse_response( $server, @_ );

                if ($result) {
                    if ( my $current_server = $self->_parse_nodes($result) ) {
                        %servers = ();
                        $cv_weak->send($current_server);
                        $cv_weak->clear_guard if $cv_weak;
                        return;
                    }
                    $self->throw( 'Internal',
                        "Couldn't extract servers from ", $result );
                }

                return if %servers;

                $cv_weak->croak(
                    $self->build_error(
                        'NoServers',
                        "Could not retrieve a list of active servers: ",
                        { servers => \@all_servers }
                    )
                );

                $cv_weak->clear_guard if $cv_weak;

            }
        );
    }

    return $cv;
}

our %Connection_Errors = map { $_ => 1 }
    ( 'Connection reset by peer', 'Connection refused', 'Broken pipe' );

#===================================
sub _parse_response {
#===================================
    my ( $self, $server, $content, $headers ) = @_;

    my ( $result, $json_error );
    eval { $result = $self->JSON->decode($content); 1 }
        or $json_error = ( $@ || 'Unknown JSON error' );

    $self->_log_result( $result || $content );

    my ( $status, $reason ) = @{$headers}{qw(Status Reason)};
    my $success = $status =~ /^2/;

    if ( $success && !$json_error ) {
        return $result;
    }

    my $error_params = {
        server      => $server,
        response    => $json_error ? $content : $result,
        status_code => $status,
        status_msg  => $reason,
    };

    my ( $error_type, $error_msg );
    if ( $success && $json_error ) {
        $error_type = 'JSON';
        $error_msg  = $json_error;
    }
    else {
        $error_type
            = $reason eq 'Connection timed out' ? 'Timeout'
            : $Connection_Errors{$reason} ? 'Connection'
            :                               'Request';
        $error_msg
            = $result && $result->{error}
            ? $result->{error}
            : $error_params->{status_msg} . ' ('
            . $error_params->{status_code} . ')';
    }

    my $error = $self->build_error( $error_type, $error_msg, $error_params );
    return ( undef, $error );
}

#===================================
sub cv { shift; ElasticSearch::AnyEvent::CondVar->new(@_) }
#===================================

#===================================
#===================================
package ElasticSearch::AnyEvent::CondVar;
#===================================
#===================================

use AnyEvent();
our @ISA = qw(AnyEvent::CondVar);

my $created = my $destroyed = 0;

#===================================
sub new {
#===================================
    my $proto = shift;
    my $class = ref $proto || $proto;
    my $self  = AnyEvent->condvar;
    bless $self, $class;
    my %params = @_;
    $created++;
    $self->{$_} = $params{$_} for keys %params;
    $self->{_guard} = [];
    return $self;
}

#===================================
sub guard {
#===================================
    my $self = shift;
    push @{ $self->{_guard} }, @_;
}

#===================================
sub clear_guard { shift->{_guard} = []; }
#===================================

#===================================
sub cb {
#===================================
    my $self = shift;
    return $self->SUPER::cb unless @_;
    my $cb = shift;

    my $cv = $self;
    Scalar::Util::weaken($cv) if $_[0];
    $self->SUPER::cb(
        sub {
            local $@;
            my $result = eval { $cv->recv };
            $cb->($result);
        }
    );
}

#===================================
sub DESTROY {
#===================================
    my $self = shift;
    $destroyed++;
    ( delete $self->{on_destroy} )->() if $self->{on_destroy};
}

#===================================
sub stats {
#===================================
    print "Created: $created\nDestroyed: $destroyed\n";
}

#===================================
#===================================
package ElasticSearch::AnyEvent::Error;
#===================================
#===================================
our @ISA = qw(ElasticSearch::Error);
@ElasticSearch::AnyEvent::Error::Internal::ISA
    = ( __PACKAGE__, 'ElasticSearch::Error::Internal' );
@ElasticSearch::AnyEvent::Error::Param::ISA
    = ( __PACKAGE__, 'ElasticSearch::Error::Param', );
@ElasticSearch::AnyEvent::Error::NoServers::ISA
    = ( __PACKAGE__, 'ElasticSearch::Error::NoServers' );
@ElasticSearch::AnyEvent::Error::Request::ISA
    = ( __PACKAGE__, 'ElasticSearch::Error::Request' );
@ElasticSearch::AnyEvent::Error::Connection::ISA
    = ( __PACKAGE__, 'ElasticSearch::Error::Connection' );
@ElasticSearch::AnyEvent::Error::Timeout::ISA
    = ( __PACKAGE__, 'ElasticSearch::Error::Timeout' );
@ElasticSearch::AnyEvent::Error::JSON::ISA
    = ( __PACKAGE__, 'ElasticSearch::Error::JSON' );

=head1 NAME

ElasticSearch::AnyEvent - An AnyEvent API for communicating with ElasticSearch

=head1 VERSION

Version 0.01, tested against ElasticSearch server version 0.9.0.

Note: This is an alpha pre-release version. The interface may change.
This module is woefully undocumented and lacking tests for the moment.

=cut

=head1 DESCRIPTION

ElasticSearch::AnyEvent adds an async interface to ElasticSearch.pm.  See
L<ElasticSearch> for more details.

=cut

=head1 SYNOPSIS

    use ElasticSearch::AnyEvent();
    my $es = ElasticSearch::AnyEvent->new(servers=>'127.0.0.1:9200');

    # Blocking

        my $cv = $es->current_server;
        print $cv->recv;

        # or

        print $es->current_server->recv

    # Callback

        my $cv = $es->current_server;
        $es->cb(sub {
            my $result = shift || die $@;
            ....
        });

    # Context

        my $cv = $es->current_server;
        undef $cv;                      # cancels

        $es->refresh_servers;           # fire-and-forget
        start_event_loop();

        {
           my $cv1 = $es->foo;
           my $cv2 = $es->foo;
           $cv2->cb(...);
        }
        # $cv1 is cancelled
        # $cv2 is backgrounded / fire-and-forget

    # Multitask:

        $es->multi_task(
            action      => 'index',
            pull_queue  => sub {
                # returns a list of HASHes to be passed to $es->$action(...)
                # can return (eg) 1000 at a time
            },
            on_success  => sub {                # optional
                my ($args,$result) = @_;
                ...
            },
            on_error    => sub {                # optional
                my ($error,$args,$queue) = @_;
                ...
            },
        )->recv || die "Error";

=head1 AUTHOR

Clinton Gormley, C<< <drtech at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Clinton Gormley.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1;
